pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java

```diff
@@ -0,0 +1,121 @@
+   *   <li>They key used when the index is registered in IndexService.</li>
+   *   <li>The internal identification in v1 files and metadata persisted on disk.</li>
+   *   <li>The default toString implementation.</li>
+   *   <li>The key that identifies the index config in the indexes section inside
+   *   {@link org.apache.pinot.spi.config.table.FieldConfig}, although specific index types may choose to read other
+   *   names (for example, <code>inverted_index</code> may read <code>inverted</code> key.</li>
+   * </ul>
+   */
+  String getId();
+
+  Class<C> getIndexConfigClass();
+
+  /**
+   * The default config when it is not explicitly defined by the user.
+   */
+  C getDefaultConfig();
+
+  C getConfig(TableConfig tableConfig, Schema schema);
+
+  /**
+   * Optional method that can be implemented to ignore the index creation.
+   *
+   * Sometimes it doesn't make sense to create an index, even when the user explicitly asked for it. For example, an
+   * inverted index shouldn't be created when the column is sorted.
+   *
+   * Apache Pinot will call this method once all index configurations have been parsed and it is included in the
+   * {@link FieldIndexConfigs} param.
+   *
+   * This method do not need to return false when the index type itself is not included in the {@link FieldIndexConfigs}
+   * param.
+   */
+  default boolean shouldBeCreated(IndexCreationContext context, FieldIndexConfigs configs) {
+    return true;
+  }
+
+  /**
+   * Returns the {@link IndexCreator} that can should be used to create an index of this type with the given context
+   * and configuration.
+   *
+   * The caller has the ownership of the creator and therefore it has to close it.
+   * @param context The object that stores all the contextual information related to the index creation. Like the
+   *                cardinality or the total number of documents.
+   * @param indexConfig The index specific configuration that should be used.
+   */
+  IC createIndexCreator(IndexCreationContext context, C indexConfig)
+      throws Exception;
+
+  /**
+   * Returns the {@link IndexReaderFactory} that should be used to return readers for this type.
+   */
+  IndexReaderFactory<IR> getReaderFactory();
+
+  /**
+   * This method is used to extract a compatible reader from a given ColumnIndexContainer.
+   *
+   * Most implementations just return {@link ColumnIndexContainer#getIndex(IndexType)}, but some may try to reuse other
+   * indexes. For example, InvertedIndexType delegates on the ForwardIndexReader when it is sorted.
+   */
+  @Nullable
+  default IR getIndexReader(ColumnIndexContainer indexContainer) {
+    throw new UnsupportedOperationException();
+  }
+
+  String getFileExtension(ColumnMetadata columnMetadata);
+
+  /**
+   * Returns whether the index is stored as a buffer or not.
+   *
+   * Most indexes are stored as a buffer, but for example TextIndexType is stored in a separate lucene file.
+   */
+  default boolean storedAsBuffer() {
```

Comment:
curious how this method is going to be used?

The StarTree index data is kept in a separate file (star_tree_index) as PinotDataBuffers. Would it break any assumption here?

Reply:
This is going to be used in `TextIndexType`. `StarTree` index is not an _index_ covered by `index-spi`.

You can see how it is being used [here](https://github.com/apache/pinot/pull/10184/files#diff-a325dc96fb2d64db293b5cb310f64351e9f988fed9230e46e4a1a587e8d53eadR170) (in case the link doesn't work: file is `pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/converter/SegmentV1V2ToV3FormatConverter.java` in https://github.com/apache/pinot/pull/10184)